hypertension/dc-7101-algorithms/py-root/lib/utils.py

```diff
@@ -1,15 +1,148 @@
+        curr_count = bucket_count[key]
+        key_and_largest_count_are_ints = (
+            largest_count_bucket != None
+            and key != "less_than_one_hundred"
+            and key != "less_than_one_sixty"
+            )
+
+        if curr_count > largest_count:
+            largest_count = curr_count
+            largest_count_bucket = key
+        if (
+            curr_count == largest_count
+            and key_and_largest_count_are_ints
+            and int(key) > int(largest_count_bucket)
+        ):
+            largest_count_bucket = key
+
+    bucketed_bp_readings = list(filter(lambda d: d[f"{type}_key"] == largest_count_bucket, bp_readings))
+    bucketed_bp_readings.sort(key=operator.itemgetter("date"))
+    most_recent_reading_of_selected_type = bucketed_bp_readings[-1][type]
+
+    return most_recent_reading_of_selected_type
+
+def bp_readings_meet_date_specs(date_of_claim, bp_readings):
+    claim_within_six_months = False
+    claim_within_one_month = False
+    date_of_claim_date = datetime.strptime(date_of_claim, "%Y-%m-%d").date()
+
+    for reading in bp_readings:
+        bp_reading_date = datetime.strptime(reading["date"], "%Y-%m-%d").date()
+        within_thirty_days = (date_of_claim_date - bp_reading_date).days <= 30
+
+        if (within_thirty_days and not claim_within_one_month):
+            claim_within_one_month = True
+        elif (within_thirty_days and claim_within_one_month):
+            claim_within_six_months = True
+        elif ((date_of_claim_date - bp_reading_date).days <= 180 and not claim_within_six_months):
+            claim_within_six_months = True
+
+    return claim_within_one_month and claim_within_six_months
+
+
+def history_of_diastolic_bp(request):
+    diastolic_history_calculation = {
+        "success": True
+    }
+    bp_readings = request["bp"]
+    bp_readings_length = len(bp_readings)
+    readings_greater_or_equal_to_one_hundred = 0
+    
+    if bp_readings_length > 0:
+        for reading in bp_readings:
+            if reading["diastolic"] >= 100:
+                readings_greater_or_equal_to_one_hundred += 1
+        diastolic_history_calculation["diastolic_bp_predominantly_100_or_more"] = True if readings_greater_or_equal_to_one_hundred / bp_readings_length >=.5 else False
+    else:
+        diastolic_history_calculation["success"] = False
+
+    return diastolic_history_calculation
+
+def sufficient_to_autopopulate (request):
+    predominance_calculation = {
+        "success": True,
+    }
+    date_of_claim = request["date_of_claim"]
+    filter_bp_readings = []
+    date_of_claim_date = datetime.strptime(date_of_claim, "%Y-%m-%d").date()
+
+    if len(request["bp"]) < 1:
+        predominance_calculation["success"] = False
+        return predominance_calculation
+
+    for reading in request["bp"]:
+        bp_reading_date = datetime.strptime(reading["date"], "%Y-%m-%d").date()
+        if bp_reading_date >= date_of_claim_date - relativedelta(years=1):
```

Comment:
should we include the same day a year ago, or not? EX: the claim date is 12/1/2021, do we allow 12/1/2020? Depending on datetime (which not using), it could be included, but it seems that just by days, we wouldn't include it, right?

Reply:
Great question. I just made a note to ask VA SMEs about it.

We will fix this as a part of https://amida.atlassian.net/browse/MCP-1045

(Resolved)